server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java

```diff
@@ -176,11 +176,17 @@ public long readBlobPreferredLength() {
 
     @Override
     public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
-        if (failIfAlreadyExists == false) {
-            deleteBlobsIgnoringIfNotExists(Collections.singletonList(blobName));
-        }
         final Path file = path.resolve(blobName);
-        writeToPath(inputStream, file, blobSize);
+        try {
+            writeToPath(inputStream, file, blobSize);
+        } catch (FileAlreadyExistsException faee) {
+            if (failIfAlreadyExists) {
+                throw faee;
+            } else {
+                deleteBlobsIgnoringIfNotExists(Collections.singletonList(blobName));
```

Comment:
Not sure why you put this in an else block, but the next call to writeToPath outside of it. I would either put both in else (preferred) or none of them